File: tests/test_post_write.py
"""Tests for hooks/scripts/post-write.py — placeholder and slop detection."""

import os
import tempfile
from conftest import run_hook, parse_hook_output


def make_write_output(file_path: str, content: str = "") -> dict:
    return {
        "tool_name": "Write",
        "tool_input": {"file_path": file_path, "content": content},
    }


class TestPlaceholderDetection:
    def test_detects_todo_in_written_file(self):
        content = "def main():\n    # TODO: implement this\n    pass\n"
        with tempfile.NamedTemporaryFile(mode="w", suffix=".py", delete=False) as f:
            f.write(content)
            f.flush()
            result = run_hook("post-write.py", make_write_output(f.name, content))
            output = parse_hook_output(result)
            assert output, "Expected JSON output for TODO detection"
            hook_out = output.get("hookSpecificOutput", {})
            # post-write.py returns a deny decision for hard placeholders
            reason = hook_out.get("permissionDecisionReason", "")
            ctx = hook_out.get("additionalContext", "")
            combined = reason + ctx
            assert "TODO" in combined.upper() or "placeholder" in combined.lower() or result.returncode == 2
        os.unlink(f.name)

    def test_clean_file_passes(self):
        content = "def add(a: int, b: int) -> int:\n    return a + b\n"
        with tempfile.NamedTemporaryFile(mode="w", suffix=".py", delete=False) as f:
            f.write(content)
            f.flush()
            result = run_hook("post-write.py", make_write_output(f.name, content))
            assert result.returncode != 2
        os.unlink(f.name)

File: tests/test-post-write.mjs
import assert from "node:assert/strict";
import { unlinkSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { describe, it } from "node:test";
import { parseHookOutput, runHook } from "./helpers.mjs";

function makeWriteOutput(filePath, content = "") {
	return { tool_name: "Write", tool_input: { file_path: filePath, content } };
}

describe("PlaceholderDetection", () => {
	it("should detect TODO in written file", () => {
		const content = "def main():\n    # TODO: implement this\n    pass\n";
		const filePath = join(tmpdir(), `test-post-write-${Date.now()}.py`);
		writeFileSync(filePath, content, "utf8");
		try {
			const result = runHook(
				"post/write-quality.mjs",
				makeWriteOutput(filePath, content),
			);
			const output = parseHookOutput(result);
			assert.ok(
				output && Object.keys(output).length > 0,
				"Expected JSON output for TODO detection",
			);
			const hookOut = output.hookSpecificOutput || {};
			const combined =
				(hookOut.permissionDecisionReason || "") +
				(hookOut.additionalContext || "");
			assert.ok(
				combined.toUpperCase().includes("TODO") ||
					combined.toLowerCase().includes("placeholder") ||
					result.status === 2,
			);
		} finally {
			unlinkSync(filePath);
		}
	});

	it("should pass clean file", () => {
		const content = "def add(a: int, b: int) -> int:\n    return a + b\n";
		const filePath = join(tmpdir(), `test-post-write-${Date.now()}.py`);
		writeFileSync(filePath, content, "utf8");
		try {
			const result = runHook(
				"post/write-quality.mjs",
				makeWriteOutput(filePath, content),
			);
			assert.notEqual(result.status, 2);
		} finally {
			unlinkSync(filePath);
		}
	});
});
